compiler/src/iree/compiler/Dialect/HAL/IR/HALOpFolders.cpp

```diff
@@ -283,6 +283,172 @@ void CommandBufferPushDescriptorSetOp::getCanonicalizationPatterns(
 
 // TODO(benvanik): fold matches that are known true based on device config.
 
+//===----------------------------------------------------------------------===//
+// hal.fence.create
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Replaces a fence with no timepoints with a null value.
+struct ElideEmptyFenceCreate : public OpRewritePattern<FenceCreateOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(FenceCreateOp op,
+                                PatternRewriter &rewriter) const override {
+    if (op.getNumOperands() != 0) return failure();
+    rewriter.replaceOpWithNewOp<IREE::Util::NullOp>(op, op.result().getType());
+    return success();
+  }
+};
+
+/// Deduplicates timepoints by taking the maximum payload value of any that
+/// share the same semaphore.
+struct DeduplicateFenceCreateTimepoints
+    : public OpRewritePattern<FenceCreateOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(FenceCreateOp op,
+                                PatternRewriter &rewriter) const override {
+    if (op.getNumOperands() == 1 + 1) return failure();  // just 1 timepoint
```

Comment:
How about `op.getSemaphores().size() == 1`?

Reply:
This was mostly just a micro-optimization - querying the total number of operands is cheaper than getting a variadic field and counting its elements (which requires looking up the segment size attr and such, though maybe not on this particular op as it has same segment sizes). Changed to do that though because this is not a hot path.